include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -0,0 +1,1043 @@
+       *
+       *  // Initialize the state of a dependent variable.
+       *  static void
+       *  dependent_variable(ADNumberType       &out,
+       *                     const ADNumberType &func);
+       *
+       * @endcode
+       * where @p ADNumberType is the auto-differentiable number type and
+       * @p ScalarType is its floating point counterpart.
+       *
+       * @author Jean-Paul Pelteret, 2017
+       */
+      template<typename ADNumberType, typename = void>
+      struct Marking;
+
+
+      /**
+       * A struct to help extract certain information associated with
+       * auto-differentiable numbers.
+       *
+       * The specializations of this class have to implement the following
+       * member data and type definitions:
+       * @code
+       *
+       *   // Extract the real scalar value.
+       *   static scalar_type
+       *   value (const ADNumberType &x);
+       *
+       *   // Extract the number of directional derivatives.
+       *   static unsigned int
+       *   n_directional_derivatives (const ADNumberType &x);
+       *
+       *   // Extract the directional derivative in the specified @p direction.
+       *   static derivative_type
+       *   directional_derivative (const ADNumberType &x,
+       *                           const unsigned int  direction);
+       *
+       * @endcode
+       * where @p ADNumberType is the auto-differentiable number type.
+       *
+       * @author Jean-Paul Pelteret, 2017
+       */
+      template<typename ADNumberType, typename = void>
+      struct ExtractData;
+
+
+      /**
+       * A struct that checks that the data expected to be stored in a specialization
+       * of the ADNumberInfoFromEnum struct has been supplied. By default it is
+       * assumed that the input type does not satisfy the necessary conditions to
+       * construct this class.
+       *
+       * @author Jean-Paul Pelteret, 2017
+       */
+      template <typename T, typename = void>
+      struct HasRequiredADInfo;
+
+
+      /**
+       * Provide a convenience function to assist in the casting of some
+       * number types to other number types. On top of the standard class
+       * definition given in @p base/numbers.h , this extension allows the
+       * conversion of automatic-differentiation numbers to generic floats.
+       *
+       * This is necessary because Adol-C doesn't provide a convenient
+       * way to convert from an @p ADNumberType to floats (@p T) other than
+       * the real-type equivalent that its associated with. For Sacado, and
+       * likely other AD number types,  the floating point value stored in
+       * an @p ADNumberType must be extracted through some function that is
+       * specific to each type of AD number. This requires some specialist
+       * intervention to get at this data.
+       *
+       * @author Jean-Paul Pelteret, 2017
+       */
+      template <typename T>
+      struct NumberType;
+
+
+      namespace
```

Comment:
Make it an internal namespace -- anonymous namespaces in header files are weird.